extension/extension.proto

```diff
@@ -0,0 +1,177 @@
+  // system should verify incoming extensions
+  rpc SetSignatureVerification(SigVerificationRequest)
+      returns (SigVerificationResponse);
+
+  // FinalizeExtensions reloads any services on the system for
+  // which extensions have been installed but which have not yet been
+  // reloaded with those extensions.
+  // Please note that this may kill the connection if the service
+  // hosting the gRPC server reloads so the server is not guaranteed
+  // to return a response here.
+  rpc FinalizeExtensions(FinalizeExtensionsRequest)
+      returns (FinalizeExtensionsResponse);
+
+  // InstallExtension installs a new extension on the system.
+  // The new extension may be transferred as part of the RPC.
+  //
+  // Expected workflows are:
+  // 1. To transfer a new extension:
+  // client ______________________________________ server
+  // TransferRequest (with extension_name set) -->
+  // TransferRequest (with contents set)       -->
+  // .. repeated TransferRequest with contents are sent
+  //  until the file is transferred ..
+  // TransferRequest (with hash set)           -->
+  // .. server verifies hash matches expected ..
+  //                                           <-- InstallExtensionResponse
+  //                             (with acknowledgement = TRANSFER_COMPLETE)
+  // STREAM END
+  //
+  // 2. To transfer and install a new extension:
+  // client ______________________________________ server
+  // TransferRequest (with extension_name set) -->
+  // TransferRequest (with contents set)       -->
+  // .. repeated TransferRequest with contents are sent
+  // until the file is transferred ..
+  // TransferRequest (with hash set)           -->
+  // .. server verifies hash matches expected ..
+  //                                           <-- InstallExtensionResponse
+  //                             (with acknowledgement = TRANSFER_COMPLETE)
+  // InstallRequest (with extension_name set)  -->
+  // (note, other fields may be set )
+  // .. server installs extension according to fields.
+  //                                           <-- InstallExtensionResponse
+  //                              (with acknowledgement = INSTALL_COMPLETE)
+  // STREAM END
+  //
+  // 3. To install a pre-existing extension:
+  // client ______________________________________ server
+  // InstallRequest (with extension_name set)
+  // (note, other fields may be set )          -->
+  // .. server installs extension according to fields.
+  //                                           <-- InstallExtensionResponse
+  //                              (with acknowledgement = INSTALL_COMPLETE)
+  // STREAM END
+  // Note that all installations will install the extension, but that
+  // extension will not be applied until the affected services are reloaded.
+  // To reload services after installing extensions, issue a
+  // FinalizeExtensions RPC after finishing the InstallExtension RPC.
+  rpc InstallExtension(stream InstallExtensionRequest)
```

Comment:
I think someone else observed this in another comment, but there is `gnoi.System.SetPackage` which was designed to be a "generic" means to install a package on a system. `gnoi.os.Install` was created specifically because there were a lot of semantics around NOS installation that were beyond `SetPackage`.

Here, what's the motivation that separates this RPC from `SetPackage`? Do you envisage that we deprecate `SetPackage` if this is merged?

Reply:
I'll confess to not knowing exactly what `System.SetPackage` is supposed to be used for given the current gNOI services.
from the description:
https://github.com/openconfig/gnoi/blob/f62dc1b711ef0809752792917fb556824fbcd013/system/system.proto#L53-L54
It seems to be a very generic way to download packages, where i dont think its very clear what a package is (this may be intentional - perhaps a package is just any bundle of files).  

The comment alludes to the RPC being potentially used to download bootable images, however this functionality was later implemented in the gNOI OS package, with this comment from the PR explaining how `SetPackage` is too generic to really capture the nuances with OS installation: https://github.com/openconfig/gnoi/pull/19#issuecomment-479083625
I'd argue that the same is true here with extensions. 
There's more to installing an extension than just file transfer and "activation" (which is really all that the `SetPackage` RPC affords). 

Some examples of differences here between `Extension` RPCs and `SetPackage`:
- `SetPackage` only allows for activation of a package after transfer, it has no capability to "activate" a package which is already present on the device
- `SetPackage` only has one option, either to "activate" a package or not:
https://github.com/openconfig/gnoi/blob/f62dc1b711ef0809752792917fb556824fbcd013/system/system.proto#L320-L323
There are other options which are relevant to extension installation, related to signature verification, dependency checks, whether this extension should be continuously applied after reboots. 
This is specific to installing an `extension` and not just something that would be relevant to any generic package.
- `Extension` allows for the installation/activation workflow to be spread over multiple RPCs (`InstallExtension` and `FinalizeExtensions`, which gives an opportunity for a network operator to install multiple extensions in sequence, verifying that they all succeed before finalizing and reloading the relevant services. 


>Here, what's the motivation that separates this RPC from SetPackage? Do you envisage that we deprecate SetPackage if this is merged?

I would be ok with deprecating it, but im not sure how widely used it is or what other vendors/network operators use it for. 
Deprecating it would be a different PR/issue though, made as a followup after this PR

Reply:
The spec here doesn't seem to really capture the nuances that you mention above that mean that `SetPackage` isn't the right solution. Can I suggest that we start with a brief design document that describes what we're trying to solve here -- particularly covering the primitives that you propose so that other implementors and operators can comment on this?

Note all the things that you mention about `SetPackage` could be done by augmenting its functionality to include the specific gaps. I don't know that is the right solution, but I also don't have a good picture about what workflows you're trying to achieve here -- which the design doc would help close.

Reply:
I've added a new commit with a README.md file outlining what an extension is and how the typical extension installation workflow would proceed, please take a look.

Reply:
Hi rob, 

Can you take a look at the README.md in this PR?  
It would be good to get this PR finalized

Reply:
Thanks @robshakir and thanks to @brianneville for adding the readme, managing extensions is something lacking in gnoi.

Would it make sense to split the transfer of an extension and its installation? For example, could we use `os.SetPackage` or `file.Put` to transfer the extension to the device and then this service could be dedicated to extension management. 

Other comments:

* Could the signature verification be made part of the Installation? Given that an individual Installation can opt out of the verification it would seem to make sense to specify the verification semantics as part of the Install.
* Would it be possible to add API for listing installed and/or activated extensions?

Reply:
>Would it make sense to split the transfer of an extension and its installation? For example, could we use os.SetPackage or file.Put to transfer the extension to the device and then this service could be dedicated to extension management.

I think having a separate way to transfer extensions onto the system is reasonable here, rather than reusing `file.Put`  for the reasons in the above comment https://github.com/openconfig/gnoi/pull/160#discussion_r1472785123 (i.e.  gNSI Authz is RPC based and we should be relucant to grant someone access to "uploading any files onto the system", and it simplifies installation if only one service client is required).

Additionally, if the client uses `file.Put`, then they'd need to know where to upload the extension on the system, rather than having that be abstracted away. (i.e. they need to know the vendor-specific location of where extension files should be placed on the device).

>For example, could we use os.SetPackage or file.Put to transfer the extension to the device and then this service could be dedicated to extension management.

I'm still not in favour of using `os.SetPackage`, for the reasons given in my comments above.
I really don't think its adequate in its current form, and i don't like the idea of overhauling that single RPC to try and force all this extension management functionality (as well as any future functionality) into it.
I dont think its a good idea to have to balance "all future extension-related changes" with "maintaining backwards compatability" for `os.SetPackage`. 
To me it seems more reasonable to just leave `os.SetPackage` alone, and develop this extension service in isolation.

 > Other comments:
>Could the signature verification be made part of the Installation? Given that an individual Installation can opt out of the verification it would seem to make sense to specify the verification semantics as part of the Install.

 yep this sounds fine 

>  Would it be possible to add API for listing installed and/or activated extensions?

Would it make more sense to add this information into some YANG model (since its just pure state for the system), so gNMI clients could subscribe to it or get it? 
If you think not though i can definitely add this here, having a way to list extension information is a good idea.

Reply:
Thanks Brian,

I think it would be good to support List extensions here initially and perhaps later adding this to gNMI if it makes sense.

Do you think there should be a third oneof type in InstallExtensionRequest to Activate extensions? I think there is a distinction between installation and activation or is the intent to activate the extension during the call to FinalizeExtensions?

Reply:
Ah understood - thanks for the clarification.

@earies @aaronmillisor @harveyyoung Would you be able to take a look at this and comment whether this is compatible with your respective NOSs?

Reply:
Just trying to follow up here. Any comment on this? It'd be good if we can finalize this PR

Reply:
I am also following up here, as we'd like to make progress internally. What will it take to get this over the finish line?

@alshabib if we are unable to get further input, can we look at approving this?

Reply:
@robshakir , same question but directed to you I guess?

Reply:
@alshabib @robshakir any feedback?